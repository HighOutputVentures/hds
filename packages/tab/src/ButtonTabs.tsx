import { Tab, TabList, TabPanel, TabPanels, Tabs } from '@chakra-ui/react';
import * as React from 'react';
import { v4 as uuid } from 'uuid';

type Item = {
  label: string;
  render(): JSX.Element | null;
};

type Placement = 'center' | 'end' | 'start';

export interface ButtonTabsProps {
  items?: Item[];
  isFitted?: boolean;
  placement?: Placement;
  /**
   *
   * places panel on top
   *
   */
  isInverted?: boolean;
  preferMounted?: boolean;
}

export default function ButtonTabs({
  items = [],
  isFitted,
  placement = 'start',
  isInverted,
  preferMounted,
}: ButtonTabsProps) {
  const [selectedIndex, setSelectedIndex] = React.useState(0);

  const tablist = React.useMemo(() => {
    return (
      <TabList justifyContent={placement} data-testid="hds.button-tabs.list">
        {items?.map(({ label }) => {
          return (
            <Tab
              key={uuid()}
              color="Gray.800"
              paddingY="10px"
              paddingX="16px"
              borderStyle="solid"
              borderColor="Gray.300"
              borderTopWidth="1px"
              borderBottomWidth="1px"
              borderRightWidth="1px"
              fontSize="14px"
              lineHeight="20px"
              fontWeight="medium"
              whiteSpace="nowrap"
              _first={{
                borderLeftWidth: '1px',
                roundedLeft: '6px',
              }}
              _last={{
                roundedRight: '6px',
              }}
              _selected={{
                bgColor: 'Gray.50',
              }}
              data-testid="hds.button-tabs.tab"
            >
              {label}
            </Tab>
          );
        })}
      </TabList>
    );
  }, [items, placement]);

  const panels = React.useMemo(() => {
    return (
      <TabPanels padding={0} data-testid="hds.button-tabs.panels">
        {items.map(({ render: Component }) => (
          <TabPanel key={uuid()} padding={0} data-testid="hds.button-tabs.panels.panel">
            <Component />
          </TabPanel>
        ))}
      </TabPanels>
    );
  }, [items]);

  return (
    <Tabs
      variant="unstyled"
      width="full"
      height="full"
      margin={0}
      padding={0}
      isFitted={isFitted}
      index={selectedIndex}
      onChange={setSelectedIndex}
      {...(!preferMounted && {
        isLazy: true,
        lazyBehavior: 'unmount',
      })}
      data-testid="hds.button-tabs.tabs"
    >
      {!isInverted && (
        <React.Fragment>
          {tablist}
          {panels}
        </React.Fragment>
      )}

      {isInverted && (
        <React.Fragment>
          {panels}
          {tablist}
        </React.Fragment>
      )}
    </Tabs>
  );
}
