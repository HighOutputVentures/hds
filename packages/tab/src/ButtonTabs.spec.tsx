import renderer from 'react-test-renderer';
import ButtonTabs from './ButtonTabs';

describe('Snapshot (ButtonTabs)', () => {
  it('Should match snapshot', () => {
    const component = renderer.create(
      <ButtonTabs
        items={[
          {
            label: 'One',
            render() {
              return <div>One</div>;
            },
          },
          {
            label: 'Two',
            render() {
              return <div>Two</div>;
            },
          },
          {
            label: 'Three',
            render() {
              return <div>Three</div>;
            },
          },
        ]}
      />,
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    component.unmount();
  });
});
