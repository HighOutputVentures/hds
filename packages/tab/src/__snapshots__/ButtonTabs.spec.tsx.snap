// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot (ButtonTabs) Should match snapshot 1`] = `
<div
  className="chakra-tabs css-smr3nb"
  data-testid="hds.button-tabs.tabs"
>
  <div
    aria-orientation="horizontal"
    className="chakra-tabs__tablist css-8owaep"
    data-testid="hds.button-tabs.list"
    onKeyDown={[Function]}
    role="tablist"
  >
    <button
      aria-controls="tabs-:r0:--tabpanel--1"
      aria-selected={false}
      className="chakra-tabs__tab css-17028si"
      data-testid="hds.button-tabs.tab"
      id="tabs-:r0:--tab--1"
      onClick={[Function]}
      onFocus={[Function]}
      role="tab"
      tabIndex={-1}
      type="button"
    >
      One
    </button>
    <button
      aria-controls="tabs-:r0:--tabpanel--1"
      aria-selected={false}
      className="chakra-tabs__tab css-17028si"
      data-testid="hds.button-tabs.tab"
      id="tabs-:r0:--tab--1"
      onClick={[Function]}
      onFocus={[Function]}
      role="tab"
      tabIndex={-1}
      type="button"
    >
      Two
    </button>
    <button
      aria-controls="tabs-:r0:--tabpanel--1"
      aria-selected={false}
      className="chakra-tabs__tab css-17028si"
      data-testid="hds.button-tabs.tab"
      id="tabs-:r0:--tab--1"
      onClick={[Function]}
      onFocus={[Function]}
      role="tab"
      tabIndex={-1}
      type="button"
    >
      Three
    </button>
  </div>
  <div
    className="chakra-tabs__tab-panels css-7m8sgd"
    data-testid="hds.button-tabs.panels"
  >
    <div
      aria-labelledby="tabs-:r0:--tab-0"
      className="chakra-tabs__tab-panel css-47dblg"
      data-testid="hds.button-tabs.panels.panel"
      hidden={false}
      id="tabs-:r0:--tabpanel-0"
      role="tabpanel"
      tabIndex={0}
    >
      <div>
        One
      </div>
    </div>
    <div
      aria-labelledby="tabs-:r0:--tab-1"
      className="chakra-tabs__tab-panel css-47dblg"
      data-testid="hds.button-tabs.panels.panel"
      hidden={true}
      id="tabs-:r0:--tabpanel-1"
      role="tabpanel"
      tabIndex={0}
    />
    <div
      aria-labelledby="tabs-:r0:--tab-2"
      className="chakra-tabs__tab-panel css-47dblg"
      data-testid="hds.button-tabs.panels.panel"
      hidden={true}
      id="tabs-:r0:--tabpanel-2"
      role="tabpanel"
      tabIndex={0}
    />
  </div>
</div>
`;
